File: packages/rust/src/bootstrap/actions/choice.rs
//! CHOOSE_OPTION action implementation.

use async_trait::async_trait;
use std::sync::Arc;

use crate::error::{PluginError, PluginResult};
use crate::prompts::CHOOSE_OPTION_TEMPLATE;
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{ActionResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Action;

/// Action for choosing from available options.
pub struct ChooseOptionAction;

#[async_trait]
impl Action for ChooseOptionAction {
    fn name(&self) -> &'static str {
        "CHOOSE_OPTION"
    }

    fn similes(&self) -> &[&'static str] {
        &["SELECT_OPTION", "PICK_OPTION", "SELECT_TASK", "PICK_TASK"]
    }

    fn description(&self) -> &'static str {
        "Choose an option from available choices. Used for task selection \
         and decision-making when multiple options are presented."
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        true
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let state = state.ok_or_else(|| {
            PluginError::StateRequired("State is required for CHOOSE_OPTION action".to_string())
        })?;

        // Compose state
        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES", "ACTION_STATE"])
            .await?;

        // Get template
        let template = runtime
            .character()
            .templates
            .get("chooseOptionTemplate")
            .map(|s| s.as_str())
            .unwrap_or(CHOOSE_OPTION_TEMPLATE);

        let prompt = runtime.compose_prompt(&composed_state, template);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse response XML".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let selected_id = parsed.get("selected_id").cloned().unwrap_or_default();

        if selected_id.is_empty() {
            return Err(PluginError::InvalidInput("No option selected".to_string()));
        }

        Ok(ActionResult::success(format!("Selected option: {}", selected_id))
            .with_value("success", true)
            .with_value("selectedId", selected_id.clone())
            .with_value("thought", thought.clone())
            .with_data("actionName", "CHOOSE_OPTION")
            .with_data("selectedId", selected_id)
            .with_data("thought", thought))
    }
}

File: packages/python/elizaos/bootstrap/actions/choice.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.bootstrap.utils.xml import parse_key_value_xml
from elizaos.prompts import CHOOSE_OPTION_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class ChooseOptionAction:
    name: str = "CHOOSE_OPTION"
    similes: list[str] = field(
        default_factory=lambda: ["SELECT_OPTION", "PICK_OPTION", "SELECT_TASK", "PICK_TASK"]
    )
    description: str = (
        "Choose an option from available choices. Used for task selection "
        "and decision-making when multiple options are presented."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        if message.content and message.content.options:
            return len(message.content.options) > 0
        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for CHOOSE_OPTION action")

        available_options: list[dict[str, str]] = []
        if message.content and message.content.options:
            available_options = message.content.options

        if not available_options:
            return ActionResult(
                text="No options available to choose from",
                values={"success": False, "error": "no_options"},
                data={"actionName": "CHOOSE_OPTION"},
                success=False,
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ACTION_STATE"])

        options_context = "\n".join(
            f"- [{opt.get('id', idx)}] {opt.get('label', '')}: {opt.get('description', '')}"
            for idx, opt in enumerate(available_options)
        )

        template = (
            runtime.character.templates.get("chooseOptionTemplate")
            if runtime.character.templates and "chooseOptionTemplate" in runtime.character.templates
            else CHOOSE_OPTION_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)
        prompt = prompt.replace("{{options}}", options_context)

        response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response_text)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response")

        thought = str(parsed_xml.get("thought", ""))
        selected_id = str(parsed_xml.get("selected_id", ""))

        if not selected_id:
            raise ValueError("No option selected")

        selected_option = next(
            (opt for opt in available_options if str(opt.get("id", "")) == selected_id),
            None,
        )

        if selected_option is None:
            raise ValueError(f"Selected option ID '{selected_id}' not found")

        response_content = Content(
            thought=thought,
            text=f"Selected option: {selected_option.get('label', selected_id)}",
            actions=["CHOOSE_OPTION"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Selected option: {selected_option.get('label', selected_id)}",
            values={
                "success": True,
                "selectedId": selected_id,
                "selectedLabel": selected_option.get("label", ""),
                "thought": thought,
            },
            data={
                "actionName": "CHOOSE_OPTION",
                "selectedOption": selected_option,
                "thought": thought,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Which task should we work on first?"),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll select the most urgent task.",
                        actions=["CHOOSE_OPTION"],
                    ),
                ),
            ],
        ]


choose_option_action = Action(
    name=ChooseOptionAction.name,
    similes=ChooseOptionAction().similes,
    description=ChooseOptionAction.description,
    validate=ChooseOptionAction().validate,
    handler=ChooseOptionAction().handler,
    examples=ChooseOptionAction().examples,
)
